lib/timeline/timeline_controller.dart

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'package:vm_service_lib/vm_service_lib.dart' hide TimelineEvent;
+
+import '../globals.dart';
+import 'timeline_protocol.dart';
+
+/// This class must not have direct dependencies on dart:html.
+///
+/// This allows tests of the complicated logic in this class to run on the VM
+/// and will help simplify porting this code to work with Hummingbird.
+class TimelineController {
+  // Frame data.
+  static const int maxFrames = 120;
+  List<TimelineFrame> frames = <TimelineFrame>[];
+  final StreamController<TimelineFrame> _frameAddedController =
+      StreamController<TimelineFrame>.broadcast();
+  Stream<TimelineFrame> get onFrameAdded => _frameAddedController.stream;
+  final StreamController<Null> _framesClearedController =
+      StreamController<Null>.broadcast();
+  Stream<Null> get onFramesCleared => _framesClearedController.stream;
+
+  // Timeline data.
+  List<TimelineThreadEvent> dartEvents = <TimelineThreadEvent>[];
```

Comment:
same here make everything that can be final, final

Reply:
done.